[1단계 - 자동차 경주 구현] 정(정진우) 미션 제출합니다

bugoverdose opened the pull request:
안녕하세요! 정이라고 합니다.

뭔가 다사다난했는데 시간 내에 간신히 제출합니다.

1. 초반 커밋들을 보면 TDD에 대해 강박관념이 있어서 private 메서드까지 Reflect를 통해 억지로 구현했는데, 지금은 전부 없애고 테스트가 필요하다고 여겨지면 public 메서드를 활용하여 간접 테스트를 하는걸로 전부 수정했습니다.

2. 또한 의도적으로 실패하는 테스트 코드(RED)와 구현 코드(GREEN)를 쪼개서 커밋했는데, 실패하는 코드는 커밋하지 말라고 오늘 배웠네요. 이미 커밋이 너무 쌓여서 그 부분은 수정 못했습니다.

뭔가 별도로 질문할 내용은 안 떠오르네요. 

미리 감사드립니다.

bugoverdose committed:
test(study): check split method in String class

bugoverdose committed:
test(study): check substring method in String class

bugoverdose committed:
test(study): check charAt method in String class

bugoverdose committed:
refactor(study): add Displayname for each test

bugoverdose committed:
test(study): check charAt throws IndexOutOfBoundsException

bugoverdose committed:
test(study): add assertThatExceptionOfType as example code

bugoverdose committed:
test(study): check size method in Set class

bugoverdose committed:
test(study): check contains method in Set class with ParameterizedTest

bugoverdose committed:
test(study): check contains method in Set class with CsvSource as arguments

bugoverdose committed:
style(study): format StringTest file

kbsat committed:
test(study): check StringIndexOutOfBoundsException extends IndexOutOfBoundsException

kbsat committed:
test(calculator): setup test for splitAndSum static method

kbsat committed:
feat(calculator): implements null or empty argument for splitAndSum method

kbsat committed:
test(calculator): add DisplayName for each test case

kbsat committed:
test(study): add name for each ParameterizedTest

kbsat committed:
feat(calculator): use multiple separators using regex for splitAndSum method

kbsat committed:
test(calculator): add custom separtor test case for splitAndSum method

kbsat committed:
feat(calculator): use custom separator if exists in splitAndSum argument

bugoverdose committed:
test(calculator): setup Exception cases for splitAndSum method

bugoverdose committed:
feat(calculator): splitAndSum throws Exception for negative argument

bugoverdose committed:
refactor(calculator): implement componse method pattern on splitAndSum

bugoverdose committed:
refactor(calculator): change split delimiter regex format

bugoverdose committed:
docs(README): setup feature list for racingcar mission

bugoverdose committed:
test(Car): setup test cases for Car

bugoverdose committed:
feat(Car): implement Car domain

bugoverdose committed:
test(Car): fix go method test case via reflection

bugoverdose committed:
test(RandomUtils): setup generateNumber method to return integer between 0 and 9

bugoverdose committed:
feat(RandomUtils): implement generateNumber method

bugoverdose committed:
feat(view): request user input for car names

bugoverdose committed:
test(study): check split method regards continuous separators as empty string

bugoverdose committed:
test(GameManager): setup splitUserInputByComma to return split String

bugoverdose committed:
feat(GameManager): split car names input by comma

bugoverdose committed:
test(GameManager): setup initCars to generate car list

bugoverdose committed:
feat(GameManager): implement initCars method

bugoverdose committed:
test(ValidatorUtils): setup validatePositiveInt to throw Exception on non-positive

bugoverdose committed:
feat(ValidatorUtils): implement validatePositiveInt

bugoverdose committed:
"test(GameManager): setup getValidateTotalRounds pass and exception cases"

bugoverdose committed:
feat(GameManager): request user input and init totalRounds with validation

bugoverdose committed:
feat(GameManager): print user input request message

bugoverdose committed:
feat(view): static methods for requesting user input

bugoverdose committed:
feat(view): implement OutputView static methods for game logic messages

bugoverdose committed:
docs(README): update feature list

bugoverdose committed:
refactor(racingcar): change package structure and match test to main

bugoverdose committed:
test(GameManager): setup getValidateCarNames test cases on pass and exception

bugoverdose committed:
feat(GameManager): implement getValidateCarNames method

bugoverdose committed:
test(ValidatorUtils): setup validations for carNames input

bugoverdose committed:
feat(ValidatorUtils): implement validations for carNames input

bugoverdose committed:
refactor(GameManager): refactor getValidateCarNames method with ValidatorUtils

bugoverdose committed:
feat(GameManager): play all rounds and print round results

bugoverdose committed:
feat(GameManager): get and print winners as game result

bugoverdose committed:
feat(CarRepository): implement car repository for car instances

bugoverdose committed:
refactor(GameManager): replace cars with car repository

bugoverdose committed:
refactor(GameManager): separate input view logic from GameManager

bugoverdose committed:
refactor(view): implement custom request user input methods

bugoverdose committed:
refactor(racingcar): separate totalRounds validation from GameManager

bugoverdose committed:
refactor(racingcar): separate car names validation from GameManager

bugoverdose committed:
refactor(racingcar): separate GameManager to Game domain and GameController

bugoverdose committed:
refactor(Game): move playAllRounds method to GameController

bugoverdose committed:
refactor(Game): use stream for getWinners to reduce depth

bugoverdose committed:
refactor(GameController): divide run logic to three steps

bugoverdose committed:
refactor(Car): canGo method to go or not

bugoverdose committed:
test(Car): remove unnecessary test on private method

bugoverdose committed:
refactor(constants): replace magic numbers and string values with constants

bugoverdose committed:
style(test): add displaynames for Game and ValidatorUtils test cases

bugoverdose committed:
refactor(calculator): remove hashmap

ddu0422 commented on README.md:
```diff
@@ -1,7 +1,27 @@
+- [x] 경주할 자동차들의 이름을 사용자로부터 입력받는다.
 
-- [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+  - [x] 자동차 이름은 쉼표(,)를 기준으로 구분하여 받는다.
+  - [x] 자동차명으로 빈문자열 혹은 공백만 입력되면 예외가 발생한다.
+  - [x] 자동차명이 중복될 경우 예외가 발생한다.
+  - 자동차명이 1대만 입력되어도 예외가 발생하지 않는다.
+  - [x] 자동차 이름이 5글자를 넘어가면 예외가 발생한다.
+  
+- [x] 사용자로부터 전체 라운드 횟수로 양의 정수를 입력받는다.
+  
+  - [x] 횟수로 정수 이외의 값이 입력되면 예외가 발생한다.
+  - [x] 횟수로 0이 입력되면 예외가 발생한다.
+  - [x] 횟수로 음수가 입력되면 예외가 발생한다.
+
+- [x] 각 라운드마다 각 자동차는 전진 혹은 정지한다.
+
+  - [x] 각 자동차는 0~9 사이의 숫자를 임의로 생성하고, 4 이상이면 전진한다.
+  - [x] 각 라운드 종료시, 모든 자동차의 현재 위치를 출력한다.
+
+- [x] 전체 라운드 종료 후 우승자를 출력한다.
+
+  - [x] 우승자는 가장 멀리 전진한 자동차가 된다.
+  - [x] 우승자가 두 명 이상이어도 예외가 발생하지 않는다.
```
요구사항 정리 👍🏻

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,78 @@
+
+public class StringCalculator {
+
+    public static final String CUSTOM_SEPARATOR_REGEX = "//(.)\n(.*)";
```
해당 클래스에서만 사용하는데 public으로 선언하신 이유가 있을까요?

bugoverdose replied:
아 실수했는데 리팩토링할 때 발견하지 못했었네요. 얘도 private으로 고쳐야겠네요.

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static String extractOperands(String rawInput) {
+        Matcher matcher = Pattern.compile(CUSTOM_SEPARATOR_REGEX).matcher(rawInput);
+        if (customSeparatorExists(matcher)) {
```
해당 메서드가 실행될 때마다 Pattern 객체가 생성되는 데 비용이 커지지 않을까요?
static으로 선언하고 재사용하는 건 어떨까요?

bugoverdose replied:
넵 그러는게 좋은거 같아요.

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,78 @@
+    private static void validateNotNegative(int integer) {
+        if (integer < 0) {
+            throw new RuntimeException("음수를 입력하면 안 됩니다.");
+        }
```
number와 같은 네이밍은 어떨까요?

bugoverdose replied:
아 네 저희가 너무 깊게 생각한 것 같네요.

ddu0422 commented on src/main/java/racingcar/constants/OutputMessages.java:
```diff
@@ -0,0 +1,18 @@
+    public static final String CAR_INFO_DELIMITER = " : ";
+    public static final String DISTANCE_SYMBOL = "-";
+    public static final String WINNERS_JOINER_DELIMITER = ", ";
+    public static final String ANNOUNCE_GAME_WINNER_TEXT = "가 최종 우승했습니다.";
```
상수도 각 책임에 맞게 각 클래스에 정의하는 건 어떨까요?

bugoverdose replied:
괜찮은것 같아요. 그런데 SystemConstants 클래스의 COMMA처럼 여러곳에서 중복으로 사용되는 상수의 경우에는 지금처럼 별도로 클래스를 두는게 나을려나요?

ddu0422 replied:
저 같은 경우는 각각의 책임이 있다고 생각해서 별도 클래스로 두기보다는 관련 클래스에서 사용할 것 같아요 :)
COMMA라는 변수명보다는 해당 객체에서 사용하는 의미가 따로 존재할 수 있을 것 같기 때문이죠.

고민을 해봐도 다른 클래스로 분리하는 게 의미가 더 크다고 생각하면 다른 클래스로 두셔도 됩니다 !_!..
단순히 제 생각이니 정도 고민하며 개발 철학을 잡아가면 좋을 것 같아요 :)

bugoverdose replied:
넵 케이스별로 고민하면서 진행해볼께요.

ddu0422 commented on src/main/java/racingcar/constants/OutputMessages.java:
```diff
@@ -0,0 +1,18 @@
+    public static final String REQUEST_CAR_NAME_INPUT_MESSAGE = "경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).";
+    public static final String REQUEST_TOTAL_ROUNDS_INPUT_MESSAGE = "시도할 회수는 몇회인가요?";
+
+    public static final String ROUND_RESULT_TEXT = "\n실행 결과";
```
\n보단 System.lineSeparator(); 을 사용하는 건 어떨까요? OS마다 개행이 달라질 수 있어요 :)

bugoverdose replied:
아 네 저희가 개행을 좀 신경 안 쓴 것 같네요.

ddu0422 commented on src/main/java/racingcar/constants/SystemConstants.java:
```diff
@@ -0,0 +1,14 @@
+    public static final int ZERO = 0;
+    public static final int TEN = 10;
+    
+    public static final String COMMA = ",";
```
다음과 같은 변수명은 의미가 불분명하지 않을까요? 해당 숫자, 문자를 영어로 바꾼 것 의미만 있지 않을까요?
더 좋은 네이밍은 없는 지 고민해보면 좋을 것 같아요 :)

bugoverdose replied:
넵넵

ddu0422 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,38 @@
+
+    public int getPosition() {
+        return position;
+    }
```
equals && hashcode & toString을 재정의해보는 건 어떨까요?
재정의하는 이유는 무엇인지 고민해보면 좋을 것 같아요 :)

bugoverdose replied:
toString으로 OutputView에 있는 메서드를 리팩토링 및 수정했고,
중복된 이름의 Car 인스턴스가 생성되었는지를 Set으로 확인하는 validation 메서드를 위해 equals & hashCode를 구현했고 테스트 케이스도 몇 개 추가했습니다.

ddu0422 commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,59 @@
+        int maxPosition = getMaxPosition();
+
+        return getCars().stream()
+                .filter((car) -> car.getPosition() == maxPosition)
```
자동차의 위치를 가져와서 비교하는 것이 아닌 메시지를 보내는 방식으로 변경해보면 어떨까요?

bugoverdose replied:
네 car 인스턴스의 isWinner 메서드를 호출하는 방식으로 수정해봤습니다.

ddu0422 commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,59 @@
+    private int getMaxPosition() {
+        return getCars().stream()
+                .mapToInt(Car::getPosition)
+                .max().orElseThrow(null);
```
만약 빈 리스트가 주어지면 Null Pointer Exception(NPE)이 발생할 것 같아요.
다른 에러를 만들거나 기본 값을 주는 형태로 변경하는 건 어떨까요?

bugoverdose replied:
음 비어있지 않은 배열이더라도 maxPosition은 0이 나올 수 있을테니, 기본값으로 아예 -1 같은 값을 줘봤습니다.

ddu0422 commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,59 @@
+                .max().orElseThrow(null);
+    }
+
+    public boolean isOver() {
+        return currentRound >= totalRounds;
+    }
```
간접적으로 테스트를 하고 있지만 isOver 자체도 public이니 해당 메서드만 검증하는 테스트를 작성해도 좋을 것 같아요 :)

bugoverdose replied:
일단 이 부분의 경우 매개변수를 받지 않고, Game 인스턴스 내부 값들만 사용해서 지금처럼 테스트하는 것밖에 방법이 안떠오르네요.

ddu0422 commented on src/main/java/racingcar/util/ValidatorUtils.java:
```diff
@@ -0,0 +1,59 @@
+
+    private static void validateNotBlank(String string) {
+        if (string.trim().isEmpty()) {
+            throw new RuntimeException(BLANK_NOT_ALLOWED_EXCEPTION);
```
isBlank()와 동일한 동작인데 isBlank()를 사용하는 건 어떨까요? [참고자료](https://www.baeldung.com/java-blank-empty-strings)

ddu0422 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    private static String requestUserInput() {
+        Scanner scanner = new Scanner(System.in);
```
해당 메서드가 실행될 때마다 scanner 객체가 생성되는 데 static으로 변경하여 재활용해보는 건 어떨까요?

bugoverdose replied:
넵넵

ddu0422 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+public class InputView {
+
+    public static String requestCarNameInput() {
+        System.out.println(REQUEST_CAR_NAME_INPUT_MESSAGE);
```
해당 내용도 OutputView와 관련된 내용이 아닐까요?

bugoverdose replied:
이 부분의 경우 원래 OutputView에 있던걸 여기로 옮겼던건데, 다시 분리하는 것도 좋을 것 같네요.

ddu0422 commented on src/main/java/racingcar/constants/OutputMessages.java:
```diff
@@ -0,0 +1,18 @@
+    public static final String REQUEST_TOTAL_ROUNDS_INPUT_MESSAGE = "시도할 회수는 몇회인가요?";
+
+    public static final String ROUND_RESULT_TEXT = "\n실행 결과";
+    public static final String EMPTY_LINE = "";
```
사용하는 곳을 확인해야 EMPTY_LINE임을 확인할 수 있을 것 같아요.
변수명만 봤을 때는 의미하는 바가 다르지 않을까요?

ddu0422 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -0,0 +1,55 @@
+            sb.append(DISTANCE_SYMBOL);
+        }
+
+        print(sb.toString());
```
| 의식적인 연습 - 2 객체지향 생활 체조 원칙 규칙 5: 줄여쓰지 않는다(축약 금지).
를 참고하여 변수명을 줄이지 않는 연습을 해보는 건 어떨까요? :)

ddu0422 commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,96 @@
+
+    private final String PARAMETERIZED_TEST_DISPLAY_FORMAT =
+            DISPLAY_NAME_PLACEHOLDER + " ["
+                    + ARGUMENTS_PLACEHOLDER + "]";
```
static 키워드를 빠뜨리셨네요!

bugoverdose replied:
아 혹시 테스트 클래스의 상수들에는 전부 static을 넣는게 좋을까요?

ddu0422 replied:
저는 테스트 클래스도 프로덕션 클래스와 동일한 원칙을 가져가는 편입니다.
그 이유는 원칙이 다르다면 학습하기도 힘들고 프로덕션 코드와 테스트 코드를 봤을 때 누군가에게는 익숙치 못한 경험이 될 수도 있을 것 같기때문입니다 :)

그래서 상수이기 때문에 `private static final` 을 추천드렸어요 :)

bugoverdose replied:
네 일관성 있는게 좋은거 같네요

ddu0422 commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,96 @@
+    @ParameterizedTest(name = PARAMETERIZED_TEST_DISPLAY_FORMAT)
+    @ValueSource(strings = {"A", "A,1", "가:2"})
+    void splitAndSum_nonIntegerThrowsException(String input) {
+        assertThatThrownBy(() -> StringCalculator.splitAndSum(input))
+                .isInstanceOf(RuntimeException.class);
```
indent가 8인 부분이 있는데 의도한 부분이실까요??

bugoverdose replied:
아마 이건 option+command+L 단축키로 인덴트 정렬했더니 이렇게 된걸로 기억해요. 눈에 거슬리게 정렬되는 경우가 종종 보였는데, 안하는 것보다는 하는게 나은 것 같아서 모든 파일들에 단축키를 적용했습니다.

ddu0422 replied:
맥을 사용하시는군요!
Preferences (command + , ) -> Editor -> Java -> Tabs and Indents -> Continuation indent 값이 8일까요?
8이라면 4로 변경해서 해결할 수 있습니다 😄 

![indent](https://user-images.githubusercontent.com/43198414/153715082-7a45ab5e-33eb-4bf1-a7bf-0adf14a16161.png)

bugoverdose replied:
감사합니다. 변경했습니다.

ddu0422 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,57 @@
+            DISPLAY_NAME_PLACEHOLDER + " ["
+                    + ARGUMENTS_PLACEHOLDER + "]";
+
+    Car car;
```
해당 클래스에서만 사용하는데 package-private로 선안한 이유가 있을까요?

bugoverdose replied:
여기도 실수를 했군요.

ddu0422 commented on src/main/java/racingcar/util/ValidatorUtils.java:
```diff
@@ -0,0 +1,59 @@
+        Set<String> carNameSet = new HashSet<>(Arrays.asList(strings));
+        if (strings.length != carNameSet.size()) {
+            throw new RuntimeException(DUPLICATE_NAME_EXCEPTION);
+        }
```
적절한 자료구조 활용이네요 💯 💯

ddu0422 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,38 @@
+    public Car(String name) {
+        this.name = name;
+        this.position = INITIAL_POSITION;
+    }
```
누군가가 자동차 객체를 생성할 때는 validation이 없어 name이 5글자를 초과해도 객체가 생성되겠군요!

bugoverdose replied:
아 validate 로직을 생성자 내부로 옮겨주는게 좋겠네요.

ddu0422 reviewed (changes requested):
안녕하세요 정! 리뷰어 미르입니다:)

첫 구현 잘 마무리 해주셨네요 💯
제가 중점적으로 본 내용은 프로그램 `동작 여부, 테스트 코드, 객체 지향 생활 체조, 읽기 좋은 코드`입니다.
현재는 궁금한 점이 없으시군요 !! 
궁금한 점이 생기면 언제든지 DM 혹은 댓글 남겨주세요 ^-^

궁금한 점이 없으시지만 도움이 되실 것 같아 내용 공유드립니다 :)

## 클래스 구현 순서

```java
public class ClassName {
    // private static final Random random = new Random();
    // 상수
    // 클래스 변수
    // 인스턴스 변수
    // 생성자
    // 팩토리 메서드
    // 일반 메서드
    // getter, setter
    // equals, hashCode, toString
}
```

긴 글 읽어주셔서 감사합니다 :)

bugoverdose committed:
refactor(calculator): use constants, fix public to private, rename parameter

bugoverdose committed:
sytle(test): continuation indent to 4 spaces

bugoverdose committed:
test: apply java code convention on test code (private, static, etc)

bugoverdose committed:
refactor(OutputView): implement printBlankLine method, private constants, replace builder with repeat

bugoverdose committed:
refactor(ValidatorUtils): use private constants and rename parameters

bugoverdose committed:
refactor(RandomUtils): rename and move TEN constant to RandomUtils class

bugoverdose committed:
test(RandomUtils): add test cases on boundary values

bugoverdose committed:
refactor(Car): private constants, initialize position outside of constructor

bugoverdose committed:
feat(Car): implement isWinner method for Car instance

bugoverdose committed:
refactor(Game): use isWinner method for getWinners

bugoverdose committed:
refactor(Game): private constants, delete unnecessary SystemConstants class

bugoverdose committed:
refactor(view): implement OutputView methods for InputView methods, remove constants package

bugoverdose committed:
feat(Car): implement toString method for OutputView

bugoverdose committed:
refactor(OutputView): modify printRoundResult method

- delete and replace printCurrentCarPosition method with toString method
- use stringBuilder
- line separator constant for builder

bugoverdose committed:
feat(Car): implement equals and hashCode

bugoverdose committed:
refactor(ValidatorUtils): modify duplicate car name input validation

- check duplicate car instances after initializing them
- validate with car instances instead of name strings
- fix and add test cases

bugoverdose committed:
refactor(ValidatorUtils): remove splitAndValidateCarNames

- move car name input validation to Car constructor
- fix and add test cases for validation methods
- add test cases for Car constructor

bugoverdose committed:
fix(Game): prevent getMaxPosition from throwing exception on empty cars list

- add test cases for getWinners
- return empty list as winners if no cars exist

ddu0422 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,72 @@
+    public String toString() {
+        return name + CAR_INFO_DELIMITER
+            + DISTANCE_SYMBOL.repeat(position);
+    }
```
현재 toString을 UI처럼 정의하셨군요! 처음에는 UI와 Domain이 분리되어 있었는데
toString을 추가하고 오히려 UI와 Domain이 엮여있는 형태가 되었어요 ㅜㅡㅜ..

toString을 정의하지 않고 생성한 Car 객체를 출력했을 때와 정의하고 생성한 Car 객체를 출력하면 어떤 값이 나올까요?
toString을 재정의 하는 목적은 https://lelecoder.com/42 를 참고하면 좋을 것 같아요 :)

bugoverdose replied:
넵. "되도록이면 toString 반환 값을 이용하여 로직을 작성하는 것은 바람직하지 않다."고 나와있군요.
현재 미션이랑 별개로 이펙티브 자바도 한번 읽어보는게 좋을 것 같네요.

ddu0422 commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,72 @@
+
+    public boolean isWinner(int winnerPosition) {
+        return position == winnerPosition;
+    }
```
메시지 정의 👍🏻 
하지만 적절한 네이밍인지 고민해보면 좋을 것 같아요.
`isWinner 메서드의 인자 값과 position이 일치하는 경우 참을 반환한다.`와 같이 테스트 케이스를 작성하셨는데 해당 메서드가 정말 우승자 인지 판단하는 걸까요? 아니면 단순히 position을 확인하는 걸까요?

ddu0422 commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,54 @@
+    private void playGameUntilEnd() {
+        OutputView.printRoundResultText();
+
+        while(!game.isOver()) {
```
해당 부분에만 while과 ( 사이에 공백이 빠졌군요 !_!

bugoverdose replied:
넵. 이것도 그렇고, 찾아보니까 [우테코 자바 스타일 가이드](https://github.com/woowacourse/woowacourse-docs/tree/master/styleguide/java)와 IntelliJ에 등록하면 되는 [xml 파일](https://github.com/woowacourse/woowacourse-docs/blob/master/styleguide/java/intellij-java-wooteco-style.xml)이 있더라고요. 모든 파일들에 이 스타일링 일괄적으로 적용해서 별도로 커밋해보겠습니다.

ddu0422 commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,62 @@
+    private int getMaxPosition() {
+        return getCars().stream()
+            .mapToInt(Car::getPosition)
+            .max().orElse(NON_EXISTING_MAX_POSITION);
```
기본값 👍🏻 👍🏻


다음 미션들을 진행하면서 예외인 상황에서 기본값으로 절대 나오지 않을 숫자와 에러를 나타낼 경우 어떤 점이 비교해보면 좋을 것 같아요 😄

ddu0422 commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,132 @@
+        assertThat(isWinner).isFalse();
+    }
+
+    @DisplayName("이름과 포지션이 동일한 Car 인스턴스는 서로 동일한 것으로 간주된다.")
+    @ParameterizedTest(name = PARAMETERIZED_TEST_DISPLAY_FORMAT)
+    @ValueSource(strings = {"ace", "pobi", "jeong"})
+    void hashCode_sameNameAndPositionMeansSameInstances(String carName) {
+        Car car1 = new Car(carName);
+        Car car2 = new Car(carName);
+
+        for (int i = 0; i < carName.length(); i++) {
+            car1.goOrNot(CAN_GO_VALUE);
+            car2.goOrNot(CAN_GO_VALUE);
+        }
+        boolean isSame = car1.hashCode() == car2.hashCode();
+
+        assertThat(isSame).isTrue();
+    }
+
+    @DisplayName("이름이 다른 Car 인스턴스는 서로 다른 것으로 간주된다.")
+    @ParameterizedTest(name = PARAMETERIZED_TEST_DISPLAY_FORMAT)
+    @ValueSource(strings = {"ace", "pobi", "jeong"})
+    void hashCode_differentNameMeansDifferentInstances(String carName) {
+        Car car1 = new Car(carName);
+        Car car2 = new Car("other");
+
+        for (int i = 0; i < carName.length(); i++) {
+            car1.goOrNot(CAN_GO_VALUE);
+            car2.goOrNot(CAN_GO_VALUE);
+        }
+        boolean isSame = car1.hashCode() == car2.hashCode();
+
+        assertThat(isSame).isFalse();
+    }
```
💯  💯

ddu0422 commented on src/main/java/racingcar/util/ValidatorUtils.java:
```diff
@@ -0,0 +1,47 @@
+    private static final String OVER_FIVE_CHARACTERS_EXCEPTION = "5글자 이하의 이름을 입력해야 합니다.";
+    private static final String DUPLICATE_CAR_EXCEPTION = "중복된 이름을 입력하면 안 됩니다.";
+
+    public static int validateAndParsePositiveInt(String string) {
```
value와 같은 네이밍은 어떨까요?_?

bugoverdose replied:
음 넵

ddu0422 reviewed (changes requested):
안녕하세요 정! 많은 고민을 하신 흔적이 보이는 코드네요 :)
특히 Commit Message가 인상적입니다 ㅎ_ㅎ

시간이 조금 있으니 마무리까지 조금만 힘내보아요 !_!...

bugoverdose committed:
fix(Car): do not use toString for OutputView

- implement formatCarPositionData for OutputView

bugoverdose committed:
refactor(Car): rename isWinner to hasSamePosition, fix use cases and test cases

bugoverdose committed:
style: apply wooteco style for all the files

bugoverdose committed:
refactor(ValidatorUtils): rename parameters

bugoverdose committed:
refactor(test): use the same paramerized test display format for all test cases

bugoverdose committed:
fix(CarRepository): remove unused clear method

- implement custom clearDatabase method for test cases

bugoverdose committed:
test(study): setup method to BeforeAll

bugoverdose committed:
refactor(RandomUtilsTest): reduce depth by maximum 1

ddu0422 reviewed (approved):
안녕하세요 정 😄 
리뷰 반영해주셔서 감사해요. 정과 이야기하며 많은 내용을 배웠답니다 :)
리뷰는 제가 이야기한 방향으로 전부 반영하지 않으셔도 괜찮아요. 정의 생각과 다른 부분이 있다면 이야기를 나누어 보고 싶으니 언제든지 말씀해주셔도 괜찮아요 😉

이번 pr은 이만 머지할게요 !_! 다음 단계도 잘해보아요 👍🏻
첫 미션 수고하셨습니다!! 궁금한 점이 있으시면 언제든지 코멘트나 DM 주세요 :)

bugoverdose commented:
@ddu0422 빠르고 섬세한 리뷰 감사했습니다!

YeonghyeonKO reviewed (commented):
LGTM
